ansible-ipi-install/roles/installer/tasks/24_rhcos_image_cache.yml

```diff
@@ -0,0 +1,124 @@
+    owner: "{{ ansible_user }}"
+    group: "{{ ansible_user }}"
+    mode: '0755'
+  when: (rhcos_openstack_stat.stat.exists == false or rhcos_openstack_stat.stat.checksum != rhcos_openstack_sha256)
+  become: yes
+
+# The following cache customizations are required for 4.3, because the installer
+# does not seem to be honoring the rhcos_image_url and cache_url values in the metal3-config
+# yaml config map (perhaps support for them was removed?).  Instead, we just put the 
+# image and its checksum where the bootstrap expects to find them by default.
+- name: Cache customizations needed for 4.3
+  block:
+    - name: Create images dir in httpd root
+      file:
+        path: "/var/www/html/images"
+        state: directory
+        owner: "{{ ansible_user }}"
+        group: "{{ ansible_user }}"
+        mode: '0755'
+      become: yes
+
+    - name: Set fact for cache image without gz extension
+      set_fact:
+        rhcos_openstack_uri_minus_gz: "{{ rhcos_openstack_uri[:-3] }}"
+
+    - name: Set fact for cache image compressed qcow2 name
+      set_fact:
+        rhcos_compressed_qcow2: "{{ rhcos_openstack_uri_minus_gz | replace('openstack','compressed') }}"
+
+    - name: Create image sub-dir in httpd root
+      file:
+        path: "/var/www/html/images/{{ rhcos_openstack_uri_minus_gz }}"
+        state: directory
+        owner: "{{ ansible_user }}"
+        group: "{{ ansible_user }}"
+        mode: '0755'
+      become: yes
+
+    # unarchive module does NOT support non-tar gz files, so we have to use shell here
+    - name: Extract {{ rhcos_openstack_uri }} into image cache sub-dir
+      shell: "gunzip -ck /var/www/html/{{ rhcos_openstack_uri }} > /var/www/html/images/{{ rhcos_openstack_uri_minus_gz }}/{{ rhcos_openstack_uri_minus_gz }}"
+      become: true
+
+    - name: Convert image into compressed qcow2
+      shell: "qemu-img convert -O qcow2 -c /var/www/html/images/{{ rhcos_openstack_uri_minus_gz }}/{{ rhcos_openstack_uri_minus_gz }} /var/www/html/images/{{ rhcos_openstack_uri_minus_gz }}/{{ rhcos_compressed_qcow2 }}"
+      become: yes
+    
+    - name: Get md5sum of new compressed qcow2 image
+      shell: "md5sum /var/www/html/images/{{ rhcos_openstack_uri_minus_gz }}/{{ rhcos_compressed_qcow2 }} | awk '{print $1}'"
+      register: rhcos_image_md5sum
+      become: yes
+
+    - name: Create md5sum file in image cache sub-dir
+      template:
+        src: rhcos-image-md5sum.j2
+        dest: "/var/www/html/images/{{ rhcos_openstack_uri_minus_gz }}/{{ rhcos_compressed_qcow2 }}.md5sum"
+        owner: "{{ ansible_user }}"
+        group: "{{ ansible_user }}"
+        mode: '0644'
+      become: yes
+  when: (release_version[0]|int == 4) and (release_version[2]|int <= 3) and ((rhcos_openstack_stat.stat.exists == false or rhcos_openstack_stat.stat.checksum != rhcos_openstack_sha256))
+
+- name: Override existing rhcos_path variable
+  set_fact:
+    rhcos_path: "http://172.22.0.1/"
```

Comment:
Does this 24_* script always run? For my FYI purposes why does it need to change from `rhcos_path: "{{ rhcos_json.json | json_query('baseURI') }}"` ?

Reply:
24_* runs when `use_rhcos_image_cache` is true.  `rhcos_path` points to the base URL where the images are located.  Initially, this value is some external URL.  It's changed here to the provisioning bridge IP for use when setting the actual cache URLs later.  An alternative would be to set a new fact that serves as the images' base URL, to differentiate it from the URL used to download the images from the Internet.

Reply:
Since in the future they will give the ability to change the IP from 172.22.0.1 (tmk), I think what we should do is likely capture the IP from the provisioning nic and then have it set that value. Where this will be tricky is at some point, we will need to differentiate between ipv4 and ipv6 and i figured i'd mention it here in case you had a bright idea for that. However, for now,  ipv4 is fine.

Reply:
> Andrew,
> 
> I think breaking this into 2 part is called for. One part that does the downloading, another that uses a cache.
> 

@sreichar Could you clarify further regarding breaking it into two parts?  Do you mean perhaps having a separate playbook/role that downloads the images?  It's true that right now a single boolean flag determines whether the images are downloaded and whether the metal3-config/install-config is set to use the pre-downloaded images.  We could easily break that up into separate flags (or sets of download and cache parameters).

> We are hoping to have a lab shared cache and can see others doing the same. So this requires checking is a caching mechanism is running and if it is not, then starting it.
> 

I think I get what you're saying here.  Perhaps another group type in the hosts called "cache" that could contain the IP of the machine that would host the httpd cache, and then a separate role to go with it.

> The using the cache will also need to the option to specify the caching source. This also means having a check that is image is available before using the cache. Obviously having default to the provision host would be the best.

Understood.  One thing to note, though, regarding checking if the image is available in the cache before using it: the bootstrap will fail-over to the standard image source (the Internet) if the supplied image source (some cache URL via metal3-config or install-config) isn't available.  In other words, the deployment already does this check for us.

Reply:
Chiming in here:

1) @abays @sreichar regarding breaking into parts, I see three flags. One flag for downloading image, 1 flag for caching, 1 flag for caching host. All flags being false by default. However, I'd see the logic as:
 IF downloading the image is set to true, then caching will be set to true. If both are true, see if a caching host has been specified, if yes, use that location otherwise create a container on the provision host to host the images. 

IF downloading the image is false and caching is set to true - look to see if there is a caching host specified. Error if no host specified and ask to either specify one or enable downloading images. 

IF caching is set to false, then downloading and cache host var auto get set to false via a set_fact and no caching is needed.

Reply:
To add:
We will need an additional group in the inventory/hosts file such as [cache-host] . We will check the length of this group in the validation.yml, if the length is 0 , we know to use the provisioning host for caching. If length > 0 then we know to use what is specified as the cache host. If length is > 1, user entered too many and ask them to only specify one cache host.

Reply:
I think @rlopez133 's suggestions are a great starting part.  Let's continue to flesh this out.